### PR TITLE
[WIP, POC] Add PolicyEnforcementMode to RBAC policy definition

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/envoyproxy/envoy/issues/3552 

Updated POC PR: https://github.com/envoyproxy/envoy/pull/3554 

Use case:  
rbac policy dark launch - test new rbac policies work as expected before it's rolled out to production, policy in dark_launch mode won't effect real users, rbac filter just log the results by applying policies and sending metrics to some metrics store/then build dashboard.

### CODE REVIEW

**Comment by mattklein123** [comment]
Nice, I like this idea. @rodaine can you take a first pass?

File: source/extensions/filters/common/rbac/engine.h [hunk 1]
```diff
@@ -25,7 +25,8 @@ class RoleBasedAccessControlEngine : public Router::RouteSpecificFilterConfig {
    *                   empty map should be used if there are no headers available.
    */
   virtual bool allowed(const Network::Connection& connection,
-                       const Envoy::Http::HeaderMap& headers) const PURE;
+                       const Envoy::Http::HeaderMap& headers,
+                       bool& darklaunch_allowed) const PURE;
```
Comment (qiwzhang) [inline source/extensions/filters/common/rbac/engine.h]: coding wise,  it is more readable to return a struct with two boolean.

Comment (diemtvu) [reply source/extensions/filters/common/rbac/engine.h]: Or return an enum. 
Alternatively, keep this function the same, but handle the result in filter only.

File: source/extensions/filters/common/rbac/engine_impl.cc [hunk 2]
```diff
@@ -26,19 +26,27 @@ RoleBasedAccessControlEngineImpl::RoleBasedAccessControlEngineImpl(
...
     return true;
   }
 
   bool matched = false;
+  bool darklaunch_matched_ = false;
   for (const auto& policy : policies_) {
     if (policy.matches(connection, headers)) {
-      matched = true;
-      break;
+      if (policy.mode() == envoy::config::rbac::v2alpha::PolicyEnforcementMode::POLICY_ENFORCEMENT_ENFORCED) {
```
Comment (diemtvu) [inline source/extensions/filters/common/rbac/engine_impl.cc]: I'm not follow this logic. Can we keep this function as is, and in the filter, simply check if poicy.isDarklaunch() (or engine.isDarklaunch()) to decide if it should reject request or simply log result?

File: source/extensions/filters/common/rbac/engine_impl.cc [hunk 3]
```diff
@@ -26,19 +26,27 @@ RoleBasedAccessControlEngineImpl::RoleBasedAccessControlEngineImpl(
...
   for (const auto& policy : policies_) {
     if (policy.matches(connection, headers)) {
-      matched = true;
-      break;
+      if (policy.mode() ==
+          envoy::config::rbac::v2alpha::PolicyEnforcementMode::POLICY_ENFORCEMENT_ENFORCED) {
+        matched = true;
+      }
+
+      darklaunch_matched_ = true;
```
Comment (rodaine) [inline source/extensions/filters/common/rbac/engine_impl.cc]: What's the intent here? We want to tell the filter that we matched a particular dark policy (or policies) correct? If so, this currently will always be true if we match any filter, and we end up hiding which policy (or policies) actually matched.

File: source/extensions/filters/common/rbac/engine.h (hunk 1 above)
Comment (quanjielin) [reply source/extensions/filters/common/rbac/engine.h]: Thanks, talked to @diemtvu offline, will use another POC PR https://github.com/envoyproxy/envoy/pull/3554, which evaluate two rules(enforcement/darklaunch) independently.

File: source/extensions/filters/http/rbac/rbac_filter.cc [hunk 4]
```diff
@@ -44,7 +44,16 @@ Http::FilterHeadersStatus RoleBasedAccessControlFilter::decodeHeaders(Http::Head
...
       config_->engine(callbacks_->route());
 
-  if (engine.allowed(*callbacks_->connection(), headers)) {
+  bool darklaunch_allowed;
+  bool enforcement_allowed = engine.allowed(*callbacks_->connection(), headers, darklaunch_allowed);
+
+  if (darklaunch_allowed) {
+    config_->stats().darklaunch_allowed_.inc();
+  } else {
+    config_->stats().darklaunch_denied_.inc();
```
Comment (rodaine) [inline source/extensions/filters/http/rbac/rbac_filter.cc]: For both the existing stats as well as these dark ones, I feel like we should apply tags with the matching policy name so we can identify the source of the allow/deny.

File: api/envoy/config/rbac/v2alpha/rbac.proto [hunk 5]
```diff
@@ -67,6 +67,14 @@ message RBAC {
   map<string, Policy> policies = 2;
 }
 
+// Policy enforcement mode, used to verify a new policy work as expected before rolling to prod.
+// RBAC filter results running from policies in dark launch mode, and discards result before return
+// to the user.
+enum PolicyEnforcementMode {
+  POLICY_ENFORCEMENT_ENFORCED = 0;
+  POLICY_ENFORCEMENT_DARKLAUNCH = 1;
+}
```
[the remark this answers is not in the document]
Comment (quanjielin) [reply api/envoy/config/rbac/v2alpha/rbac.proto]: removed this from proto in [2nd POC](https://github.com/envoyproxy/envoy/pull/3554)

File: source/extensions/filters/common/rbac/engine_impl.cc [hunk 6]
```diff
@@ -26,19 +26,27 @@ RoleBasedAccessControlEngineImpl::RoleBasedAccessControlEngineImpl(
...
   if (engine_disabled_) {
     return true;
   }
 
   bool matched = false;
+  bool darklaunch_matched_ = false;
   for (const auto& policy : policies_) {
     if (policy.matches(connection, headers)) {
-      matched = true;
-      break;
```
[the remark this answers is not in the document]
Comment (quanjielin) [reply source/extensions/filters/common/rbac/engine_impl.cc]: separated rules(enforce/darklaunch) approach in 2nd POC solves this issue

File: source/extensions/filters/http/rbac/rbac_filter.cc (hunk 4 above)
Comment (quanjielin) [reply source/extensions/filters/http/rbac/rbac_filter.cc]: agree, will put this in separated PR about what attributes need to be collected and how to pass data to metrics storage(pass data cross-filters, etc)

**Comment by quanjielin** [comment]
Close this PR since it's replaced by https://github.com/envoyproxy/envoy/pull/3554

**Pull request closed without merging**